### ISSUE TITLE
Eclipse, bnd & Concurrency

### ISSUE THREAD

**Issue by pkriens** [bndtools/bnd#4575]
There is a PR to refresh the workspace. While working on this PR, it became clear that the current Eclipse interface for bnd is a mess. We've always had the problem that we did not really know Eclipse and struggled to get things to work. Neil wrote the first incarnation of the builder, we rewrote it during a bndtools hackaton, BJ made serious improvements several times. However, as member of this group I think I can say that none of use really knew what we were doing. The result is a mess. Most of the direct problems have found their band-aids but problems are prevalent when the workspace undergoes large changes on disk.

* No rules about locking and what is thread safe to call
* No clear rule when to refresh what resource
* Synchronous inline actions, for example the builder calls things that should work from listeners
* Mostly ignoring locking
* Not taking advantage of some of the (newer?) Eclipse facilities
* We do not take advantage of parallel builds

This is an attempt to setup some rules before we start programming them.

## Concurrency

To my great surprise, parallel gradle tends to work very well. Surprise, because bnd is not designed to run concurrently. It likely works so well because Gradle builds up the Workspace and Project objects ahead of any concurrent actions and this structure is never changed during the run. In Eclipse, the structure does change and this makes it necessary to lock. We've chosen for a single lock in Central to execute all actions. A few years ago, the Workspace got a `readLocked` and `writeLocked` method that takes a lambda. So far, these have only be used by the Remote Workspace.

Looking at the code, I think we have different types of bnd code:

* **Repositories** – The repository handling must be completely thread safe and not require any global bnd locks.
* **Workspace** – Actions that change the workspace (refresh bnd files, adding/removing projects, etc.) need a write lock on the workspace. 
* **Project** – Most actions on a project could use a read lock on the workspace. The exception is refresh, it should hold a write lock on the workspace.
* **Analyzers/Builders/Launchers** – All this code should not be safe to use concurrently. Some of these classes are created via a Workspace or Project and hold a reference via the Processor. We might consider creating copies of the parent since a refresh of the parent invalidates the properties the object is based on. Open for discussion, could we create some copy on write model?

## Workspace Access

A significant problem in the current design is the access to the workspace. There is a tendency in the workspace to want to use the Workspace synchronously. However, a Workspace can take a very long time to initialize, minutes. It is therefore _really bad_ practice to call for the workspace and start using it without considering the delay. A solution that is partially in place is to only access the workspace through a Promise and execute the active code the success callback. Since this is a very common action and requires copious amount of code to handle locking, refreshing, and switching workspace, we probably should provide central support. Additionally, there is the alternative usage of Bndtools by m2e. In m2e there is no single workspace and this complicates many issues. (Someone else could fill in the details.)

### Refresh & Resources

The primary need for concurrency is channeled through the refresh. Eclipse maintains an internal resource tree that represents the file system, among other things. Originally, this tree was only updated during a refresh action that locked the Eclipse workspace. However, there are now file system watchers that refresh resources anytime. We are also particularly bad in refreshing whenever we feel like it. Worse,  we sometimes refresh while holding the bnd lock. 

We should make the resource model more central. The quickfix work was for me an example how you could save a lot of code if you allowed Eclipse to do its work in a natural way. This means that the Workspace and  Project objects must _only_ be bnd refreshed when the Eclipse resource signals a change. I've seen code that refreshes a Processor because it knows the underlying resource was changed, this is the wrong way around. There should only be one place where we refresh these objects. We will then need a listener for when Workspace & Project under go state changes. There are already some listeners in use but we should design a clean new interface based on our current knowledge. These listeners should interact in a well defined way with the Workspace locks.

When we take the resource model more serious, we should also refresh projects based on disk. The Workspace model is not always loved but we're often not taking advantage that it actually has a 1:1 representation with the file system. Many Eclipse haters list the disconnect between the file system and the resource tree as a primary reason for the their disgust. The Bndtools Explorer already has some nice features leveraging the 1:1 relation, we could do more. One of the things is to ensure the Eclipse Workspace is perfectly aligned with the file system at all times.

### Bndtools Builder

The Bndtools Builder has crammed more and more features in the build process, while having the bnd lock:

* Dependency handling
* Classpath construction
* Decorating the UI (component markers, package state, errors)
* Workspace Repository index handling
* Refreshing created artifacts
* Refreshing the current workspace when this was cnf (!)
* Working sets calculation

All these elements must be removed from the Builder (and can with newer features in Eclipse) and handled through Workspace and Project listeners. The Bndtools Builder should _only_ obtain a readlock on the workspace, build the project, and then outside the lock refresh any artifacts.

In the current builder we go out of our way to detect changes in the workspace project list and then program the dependencies and class path container. We actually sometimes need to postpone a build because other projects need to be build first. Eclipse nowadays has a org.eclipse.core.resources.builders feature called _dynamic references_. This allows a separate class to handle the dynamic dependencies _before_ the builder is invoked. Since we know the exact dependency hierarchy in the Workspace, we can guide the overall builder to call the Bndtools Builders in the proper order. Since the builder then is never called out of order, we can limit the class path container initialization when the workspace changes or a project generates new artifacts. (I've done experiments and it seems to make a significant difference in the build.)

A complicating issue is the generate phase. Since this is run _before_ the compiler runs, it is implemented as a compiler assistant. After some experience, I wonder if this is the right approach. It might be better to create a second BndGenerate builder that runs before the Java compiler. This should be discussed.

### Eclipse Parallel Build Facilities

Eclipse has facilities to run the build parallel. For the longer term survival of the Eclipse plugin it is necessary to support parallel building. That said, the support is badly documented and not well supported. The support for parallel Java compilation is actually experimental and there seems to be nobody working in moving this into mainstream.  Parallel support cannot be activated currently from the preferences UI. Experiments do show a speedup but this is likely severely restricted in our current severe and leaky locking model. 

The Bndtools builder must support parallel building. To enable this is quite straightforward. Each builder provides the overal builder with a lock (a resource). We currently use the default, which is the workspace. Returning null will not lock the tree. Experiments show that null works but the implications are not completely clear.

### Consistent Jobs Usage

Last but not least, Eclipse Jobs and locking. And it is not until recently that it became clear to me that an Eclipse Rule is generally a resource. Each Job in Eclipse has a set of 'rules'. As far as I can see these are generally Resources and are not supposed to overlap for simultaneously running jobs.  Although this works for exclusion, there seems to be no mechanism to schedule a job _after_ another job.

The primary reasons we need background jobs are:

* Refreshing – When we create or delete new artifacts, they need to be refreshed
* Long term actions – Resolving, releasing, launching, etc. require a job
* Repository views – The repository views need the full initialization of the workspace; this can be a very lengthy process

We need to verify that every job is using the proper, minimal lock. More?

**Comment by chrisrueger** [comment]
I cannot provide much at the moment other than that I really appreciate your write up. Since we moved to bndtools beginning of this year (from PDE), we already feel some of the pain points you are mentioning (long builds, Eclipse hanging because it is doing stuff... strange error popups when switching git branches, long rebuilds, build not refreshing when underling file was changed in external editor...). (... but to be fair I want to add, that our move to bndtools was worth every minute, since it allowed us to move forward on other things much faster)

So any kinds of performance speed-ups are very welcome, to make developer experience on larger projects better. 
I am glad to help where I can but have close to zero experience with developing an Eclipse app so far.

**Comment by kriegfrj** [comment]
> I am glad to help where I can but have close to zero experience with developing an Eclipse app so far.

@chrisrueger I think that means you're on roughly the same level as the rest of us. :smile:

@pkriens, thanks for the write-up. I have the following general comments to add, plus a few more specific ones:

* Tests, tests and more tests. I said it before I started work on `biz.aQute.tester.junit-platform` - the lack of regression tests are ultimately a brake on development. Especially with something as core and as critical as workspace refreshing or `BndtoolsBuilder` - people can see that the code is all over the place and perhaps not in line with the latest Eclipse best practices, they have a strong sense that it could be refactored and significantly improved, but they are too afraid to do so because the risk of breaking some obscure edge case is high and the consequences of doing so are severe. But I don't think I need to convince anyone here of that - the main reason we don't have good tests for Bndtools was because in the beginning it was hard, and that difficulty only increases over time the larger your untested codebase becomes. Hopefully the progress that we've made in `bndtools.core.tests` will give us a good platform to start from to avoid making the same mistake again.
* We should try and move to services as much as possible. Since we moved to the "facade" pattern for the quick fix processor, it has become so much easier and faster to develop tests for it. If it's quicker to test, we'll write better and more comprehensive tests.
* I've still got my eye on #3175.
* When we're in Eclipse land, maybe we should try and use Eclipse's asynchronous job processing where appropriate rather than raw Java or OSGi asynchronous constructs. Reason being that they integrate better with the UI and provide the user with more feedback as to what is going on (and what might have hung). I notice that `PromiseFactory` is used extensively in parts of Bndtools - perhaps we could come up with a `PromiseFactory` implementation that is layered on top of Eclipse's Job API? Or would that be overkill for lots of short async tasks?

A couple of specific comments:
> The Bndtools Builder should only obtain a readlock on the workspace, build the project, and then outside the lock refresh any artifacts.

Actually, I think we can do better. The builder only needs to obtain a read lock on `cnf` and the upstream dependencies projects in the workspace. Get the corresponding `IResource` for each of these and then combine them using `org.eclipse.core.runtime.jobs.MultiRule` into a single scheduling rule.

Or are you here referring to the bnd Workspace object, rather than the Eclipse workspace?

> Refreshing the current workspace when this was cnf (!)
Yes, I thought this was odd that `CnfWatcher` was inited by the builder and part of the builder package. I also just discovered that its functionality seems to partially overlap with the `WorkspaceListener`, which is installed by `Central` once the bnd `Workspace` is created.

>  it is not until recently that it became clear to me that an Eclipse Rule is generally a resource.

Ditto! (And by recently, I mean in the last couple of days, and mostly thanks to @gamerson's tip in #4569.) It demystifies scheduling rules significantly.

> ...there seems to be no mechanism to schedule a job after another job.

You can do this using `Job.join()`. But you need to have a handle on the job that you want to schedule after, or else know its special key so that you can join by name. In the second case, I'm not sure what happens if you try to join and no such task is running - does it return immediately, or wait indefinitely?

If we layered a `PromiseFactory` implementation over the top of the Eclipse `JobManager`, then that might be a way to get such a mechanism with a simple API?

> Long term actions – Resolving, releasing, launching, etc. require a job

With launching, Eclipse sets up the job for you. I think we're already pretty well covered with launching.

**Comment by pkriens** [comment]
> Or are you here referring to the bnd Workspace object, rather than the Eclipse workspace?
Yes. I think bnd needs to be aware of the concurrency & locking. We're currently skating on thin ice since bnd was  originally not developed to run concurrently. We should interact nicely but I want to support the Intellij plugin as much as possible. So locking should be based on a bnd Workspace/Project model. 

I do agree about testing but my priority is to have a set of crisp rules. I think we're currently all over the place.

**Comment by bjhargrave** [comment]
Closing as progress has been made.
